File: Homework/Hoang_Nguyen/lesson7_2/127_wordLadder.js
var buildGraph = function (beginWord, wordList) {
  const m = beginWord.length;

  // map: wildcard -> set of words
  const wildcards = new Map();

  function getWildcard(word, i) {
    return word.slice(0, i) + '*' + word.slice(i + 1);
  }

  function addToWildcards(word) {
    for (let i = 0; i < m; i++) {
      const wildcard = getWildcard(word, i);
      if (!wildcards.has(wildcard)) {
        wildcards.set(wildcard, new Set());
      }
      wildcards.get(wildcard).add(word);
    }
  }

  // build wildcards
  addToWildcards(beginWord);
  for (let i = 0; i < wordList.length; i++) {
    addToWildcards(wordList[i]);
  }
  // console.log(wildcards)

  // build graph
  // map: word -> set of connected words
  const graph = new Map();

  function addToGraph(word) {
    for (let i = 0; i < m; i++) {
      const wildcard = getWildcard(word, i);
      const connectedWords = wildcards.get(wildcard);
      for (const connectedWord of connectedWords) {
        if (!graph.has(word)) {
          graph.set(word, new Set());
        }
        graph.get(word).add(connectedWord);
      }
    }
  }

  addToGraph(beginWord);
  for (let i = 0; i < wordList.length; i++) {
    addToGraph(wordList[i]);
  }

  return graph;
};

/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function (beginWord, endWord, wordList) {
  const graph = buildGraph(beginWord, wordList);
  // console.log(graph)

  const visited = new Set();

  const queue = [];

  visited.add(beginWord);
  for (const word of graph.get(beginWord)) {
    visited.add(word);
    queue.unshift(word);
  }

  let currentDis = 1;
  while (queue.length) {
    currentDis++;
    const currentSize = queue.length;
    for (let i = 0; i < currentSize; i++) {
      const word = queue.pop();
      if (word === endWord) {
        return currentDis;
      }
      for (const nextWord of graph.get(word)) {
        if (!visited.has(nextWord)) {
          visited.add(nextWord);
          queue.unshift(nextWord);
        }
      }
    }
  }
  return 0;
};

File: Homework/Ngan_Luu/127-word-ladder.py
# 127. https://leetcode.com/problems/word-ladder/
def checkSameChar(word1, word2):
    count=0
    for i in range(len(word1)):
        if word1[i]!=word2[i]:
            count+=1
        if count==2:
            return False
    return True
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if beginWord in wordList:
            wordList.remove(beginWord)
        queue=[beginWord]
        distance=[1]
        visited={beginWord:True}
        while queue:
            word=queue[0]
            d=distance[0]
            queue.pop(0)
            distance.pop(0)
            for w in wordList:
                if w not in visited:
                    if checkSameChar(word,w):
                        if w==endWord:
                            return d+1
                        queue.append(w)
                        distance.append(d+1)
                        visited[w]='True'
        return 0